### PR TITLE
FEAT: AWS secret extension

### LINKED ISSUE #2780: Support AWS secrets in configuring Data Prepper pipelines
**Is your feature request related to a problem? Please describe.**
In the current state, user credential variables such as username and password are hardcoded in pipelines.yaml. For better security, we could support AWS secret as an extension plugin, i.e.

data-prepper-config.yaml
```
aws:
  credentials:
    default:
       sts_role_arn: arn:aws:iam::99123456789:role/OsiDataPlaneRole
       region: us-east-1
     my-custom-role:
       sts_role_arn:
       region:
  secret: # secret extension
    my_es_secret_configuration:
      name: ...
      authentication: my-custom-role
```
Then our pipeline configuration can reference the key-value pairs stored in an AWS secret.
pipelines.yaml
```
...
sink:
  - opensearch:
      ...
      username: ${{ aws_secret:my_es_secret_configuration:USERNAME }}
      password: ${{ aws_secret:my_es_secret_configuration:PASSWORD }}
```

**Describe the solution you'd like**
We can reuse aws credential extensions when configuring the credentials for AWS secrets.

**Describe alternatives you've considered (Optional)**
Alternatively, we can make AWS secret extension independent of credentials plugin at the price of duplication in configuration:

```
aws:
  credentials:
    default:
      sts_role_arn: arn:aws:iam::99123456789:role/OsiDataPlaneRole
      region: us-east-1
    my-custom-role:
      sts_role_arn:
      region:

aws_secrets:
  secrets:
    my-secret1:
      sts_role_arn: arn:aws:iam::99123456789:role/OsiDataPlaneRole
      region: us-east-1
```

**Additional context**

### PR DESCRIPTION
**Pull request by chenqi0805**
### Description
This PR adds support for AWS secret extension plugin:

```
pipeline_extensions:
  aws:
    secrets:
      secret1:
        name: my-es-secret
        region: us-east-1
      secret2:
        name: my-es-secret-max-retries
        region: us-east-1
log-pipeline:
  workers: 1
  source:
    http:
      path: "/log/ingest"
  buffer:
    bounded_blocking:
      batch_size: ${{aws_secrets:secret1:batch_size}} # max number of records the buffer will drain for each read
  sink:
    - opensearch:
        hosts: [ "..." ]
        index: test-es7-compression
        username: ${{aws_secrets:secret1:username}}
        password: ${{aws_secrets:secret1:password}}
        flush_timeout: -1
        max_retries: ${{aws_secrets:secret2}}
```

The scope of settings that support secret reference is limited to plugins (source/buffer/sink/processor) in pipeline.YAML. The reference pattern is explained as follows:

- ${{aws_secrets:secret1:username}} means the secret value under secret1 will be parsed as json and the setting value is swapped with the one according to key `username` in the secret json
- ${{aws_secrets:secret2}} means the secret value under secret2 is parsed as a literal string and the setting value is swapped with the literal string value
 
### Issues Resolved
Resolves #2780 
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/aws-plugin/src/main/java/org/opensearch/dataprepper/plugins/aws/AwsSecretManagerConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+import java.util.Optional;
+import java.util.UUID;
+
+public class AwsSecretManagerConfiguration {
+    private static final String DEFAULT_AWS_REGION = "us-east-1";
+    private static final String AWS_IAM_ROLE = "role";
+    private static final String AWS_IAM = "iam";
+
+    @JsonProperty("name")
```
Comment (dlvenable) [inline data-prepper-plugins/aws-plugin/src/main/java/org/opensearch/dataprepper/plugins/aws/AwsSecretManagerConfiguration.java:24]: AWS SecretsManager uses the term `SecretId`, so let's name this `secret_id`.

https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_GetSecretValue.html#API_GetSecretValue_RequestSyntax

File: data-prepper-plugins/aws-plugin/src/main/java/org/opensearch/dataprepper/plugins/aws/AwsSecretManagerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                .build();
+    }
+
+    public GetSecretValueRequest createGetSecretValueRequest() {
+        return GetSecretValueRequest.builder()
+                .secretId(awsSecretName)
+                .build();
+    }
+
+    private AwsCredentialsProvider authenticateAwsConfiguration() {
```
Comment (dlvenable) [inline data-prepper-plugins/aws-plugin/src/main/java/org/opensearch/dataprepper/plugins/aws/AwsSecretManagerConfiguration.java:58]: Let's use the existing AWSCredentialsProvider as provided by the AWS Plugin.

File: data-prepper-plugins/aws-plugin/src/main/java/org/opensearch/dataprepper/plugins/aws/AwsSecretsSupplier.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                    final AwsSecretManagerConfiguration awsSecretManagerConfiguration =
+                            awsSecretManagerConfigurationMap.get(secretConfigurationId);
+                    final SecretsManagerClient secretsManagerClient = entry.getValue();
+                    final GetSecretValueRequest getSecretValueRequest = awsSecretManagerConfiguration
+                            .createGetSecretValueRequest();
+                    final GetSecretValueResponse getSecretValueResponse;
+                    try {
+                        getSecretValueResponse = secretsManagerClient.getSecretValue(getSecretValueRequest);
+                    } catch (Exception e) {
+                        throw ResourceNotFoundException.builder()
```
Comment (dlvenable) [inline data-prepper-plugins/aws-plugin/src/main/java/org/opensearch/dataprepper/plugins/aws/AwsSecretsSupplier.java:43]: Why are you throwing an AWS SDK exception here?

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/RetryConfiguration.java [hunk 4]
```diff
@@ -78,7 +78,7 @@ public static RetryConfiguration readRetryConfig(final PluginSetting pluginSetti
     if (dlqFile != null) {
       builder = builder.withDlqFile(dlqFile);
     }
-    final Integer maxRetries = (Integer) pluginSetting.getAttributeFromSettings(MAX_RETRIES);
+    final Integer maxRetries = pluginSetting.getIntegerOrDefault(MAX_RETRIES, null);
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/RetryConfiguration.java:81]: Are these part of another PR? This seems unrelated.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/RetryConfiguration.java (hunk 4 above)
Comment (chenqi0805) [reply data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/RetryConfiguration.java:81]: Casting would not work for secret value substitution especially when it is a String. But it is possible with generic type in PluginConfigValueTranslator this would not be necessary

File: data-prepper-core/build.gradle [hunk 5]
```diff
@@ -16,6 +16,7 @@ dependencies {
     implementation project(':data-prepper-expression')
     implementation project(':data-prepper-plugins:blocking-buffer')
     implementation project(':data-prepper-plugins:common')
+    implementation project(':data-prepper-plugins:aws-plugin')
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-core/build.gradle:19]: Thanks for the catch! It is unnecessary

File: data-prepper-plugins/aws-plugin/src/main/java/org/opensearch/dataprepper/plugins/aws/AwsSecretsSupplier.java (hunk 3 above)
Comment (chenqi0805) [reply data-prepper-plugins/aws-plugin/src/main/java/org/opensearch/dataprepper/plugins/aws/AwsSecretsSupplier.java:43]: I was trying to wrap it with a more explicit error message. Maybe RuntimeException is more appropriate?

File: data-prepper-plugins/aws-plugin/build.gradle [hunk 6]
```diff
@@ -16,7 +20,7 @@ jacocoTestCoverageVerification {
     violationRules {
         rule {
             limit {
-                minimum = 1.0
+                minimum = 0.98
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/aws-plugin/build.gradle:23]: There is one JsonProcessingException unreachable:

```
@Override
    public Object retrieveValue(String secretId) {
        if (!secretIdToValue.containsKey(secretId)) {
            throw new IllegalArgumentException(String.format("Unable to find secretId: %s", secretId));
        }
        try {
            final Object secretValue = secretIdToValue.get(secretId);
            return secretValue instanceof Map ? OBJECT_MAPPER.writeValueAsString(secretIdToValue.get(secretId)) :
                    secretValue;
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException(String.format("Unable to read the value under secretId: %s as string",
                    secretId));
        }
    }
```
This is about writing value back into a string when ${{aws_secrets:secretId}} is used but the actual secret Value is a valid json, which means we interpret it as literal value

File: data-prepper-plugins/aws-plugin/src/main/java/org/opensearch/dataprepper/plugins/aws/AwsSecretManagerConfiguration.java (hunk 2 above)
Comment (chenqi0805) [reply data-prepper-plugins/aws-plugin/src/main/java/org/opensearch/dataprepper/plugins/aws/AwsSecretManagerConfiguration.java:58]: https://github.com/opensearch-project/data-prepper/issues/3350

File: data-prepper-plugins/aws-plugin/build.gradle (hunk 6 above)
[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-plugins/aws-plugin/build.gradle:23]: Can you mock `secretIdToValue` such that `get(secretId)` returns non-JSON?

**Review by dlvenable — APPROVED** [review]
